File: azure-resource-abbreviations/scripts/lookup_abbreviation.py
#!/usr/bin/env python3
"""
Look up abbreviation for a specific Azure resource type.
Usage: python lookup_abbreviation.py "Virtual Machine"
"""

import sys
from pathlib import Path

# Add parent directory to path
sys.path.insert(0, str(Path(__file__).parent))

from azure_abbreviations import get_abbreviation, search_abbreviation


def main():
    if len(sys.argv) < 2:
        print("Usage: python lookup_abbreviation.py <resource_type>")
        print("\nExamples:")
        print("  python lookup_abbreviation.py 'Virtual Machine'")
        print("  python lookup_abbreviation.py 'Web App'")
        print("  python lookup_abbreviation.py 'sql'")
        sys.exit(1)
    
    resource_type = " ".join(sys.argv[1:])
    
    # Try exact match first
    abbreviation = get_abbreviation(resource_type)
    if abbreviation:
        print(f"✓ {resource_type}: {abbreviation}")
        return
    
    # Try search if no exact match
    results = search_abbreviation(resource_type)
    if results:
        print(f"Found similar resources for '{resource_type}':\n")
        for category, resources in results.items():
            print(f"[{category}]")
            for res_name, abbr in resources.items():
                print(f"  {res_name}: {abbr}")
    else:
        print(f"✗ No abbreviation found for '{resource_type}'")
        sys.exit(1)


if __name__ == "__main__":
    main()

File: azure-resource-abbreviations/scripts/lookup_abbreviation.cs
#:property PublishAot=false
#:property NoWarn=CA2266
#:include AzureResourceAbbreviationsSupport.cs

if (args.Length == 0)
{
    Console.WriteLine("Usage: dotnet run --file scripts\\lookup_abbreviation.cs -- <resource_type>");
    Console.WriteLine();
    Console.WriteLine("Examples:");
    Console.WriteLine("  dotnet run --file scripts\\lookup_abbreviation.cs -- \"Virtual Machine\"");
    Console.WriteLine("  dotnet run --file scripts\\lookup_abbreviation.cs -- \"Storage Account\"");
    Console.WriteLine("  dotnet run --file scripts\\lookup_abbreviation.cs -- \"webSitesAppService\"");
    return 1;
}

var query = string.Join(' ', args);
var exactMatch = AzureResourceAbbreviationsSupport.FindExact(query);
if (exactMatch is not null)
{
    PrintEntry(exactMatch);
    return 0;
}

var results = AzureResourceAbbreviationsSupport.Search(query);
if (results.Count == 0)
{
    Console.WriteLine($"No abbreviation found for '{query}'.");
    Console.WriteLine($"Local catalog: {AzureResourceAbbreviationsSupport.GetReferencePath()}");
    return 1;
}

Console.WriteLine($"Found similar resources for '{query}':");
Console.WriteLine();

var groupedResults = results.GroupBy(entry => entry.Category, StringComparer.Ordinal);
foreach (var categoryGroup in groupedResults)
{
    Console.WriteLine($"[{categoryGroup.Key}]");
    foreach (var entry in categoryGroup)
    {
        PrintEntry(entry, "  ");
    }

    Console.WriteLine();
}

return 0;

static void PrintEntry(AzureResourceEntry entry, string indent = "")
{
    Console.WriteLine($"{indent}Display name:    {entry.DisplayName}");
    Console.WriteLine($"{indent}Resource key:    {entry.ResourceTypeKey}");
    Console.WriteLine($"{indent}Category:        {entry.Category}");
    Console.WriteLine($"{indent}Official prefix: {entry.OfficialPrefix}");
    Console.WriteLine($"{indent}Naming token:    {entry.NamingToken}");
}
